test/indexed_test.dart

```diff
@@ -0,0 +1,176 @@
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////////
+
+import 'dart:math';
+import 'dart:typed_data';
+
+import 'package:collection/collection.dart';
+import 'package:test/test.dart';
+
+import '../lib/realm.dart';
+
+part 'indexed_test.g.dart';
+
+// Don't import our own test.dart here. It will break AOT compilation.
+// We may use AOT compilation locally to manually run the performance
+// tests in this file
+
+@RealmModel()
+class _WithIndexes {
+  @Indexed()
+  late int anInt;
+
+  @Indexed()
+  late bool aBool;
+
+  @Indexed()
+  late String string;
+
+  @Indexed()
+  late DateTime timestamp;
+
+  @Indexed()
+  late ObjectId objectId;
+
+  @Indexed()
+  late Uuid uuid;
+}
+
+@RealmModel()
+class _NoIndexes {
+  late int anInt;
+  late bool aBool;
+  late String string;
+  late DateTime timestamp;
+  late ObjectId objectId;
+  late Uuid uuid;
+}
+
+typedef QueryBuilder<T, U> = RealmResults<T> Function(U value);
+
+Future<void> main([List<String>? args]) async {
+  test('Indexed faster', () {
+    final config = Configuration.local([WithIndexes.schema, NoIndexes.schema]);
+    Realm.deleteRealm(config.path);
+    final realm = Realm(config);
+
+    const max = 100000;
+    final allIndexed = realm.all<WithIndexes>();
+    final allNotIndexed = realm.all<NoIndexes>();
+
+    expect(realm.all<WithIndexes>().length, 0);
+
+    intFactory(int i) => i.hashCode;
+    boolFactory(int i) => i % 2 == 0;
+    stringFactory(int i) => '${i.hashCode} $i';
+    timestampFactory(int i) => DateTime.fromMillisecondsSinceEpoch(i.hashCode);
+    objectIdFactory(int i) => ObjectId.fromValues(i.hashCode * 1000000, i.hashCode, i);
+    uuidFactory(int i) => Uuid.fromBytes(Uint8List(16).buffer..asByteData().setInt64(0, i.hashCode));
+
+    final indexed = List.generate(
+      max,
+      (i) => WithIndexes(
+        intFactory(i),
+        boolFactory(i),
+        stringFactory(i),
+        timestampFactory(i),
+        objectIdFactory(i),
+        uuidFactory(i),
+      ),
+    );
+    realm.write(() => realm.addAll(indexed));
+
+    expect(allIndexed.length, max);
+
+    final notIndexed = List.generate(
+      max,
+      (i) => NoIndexes(
+        intFactory(i),
+        boolFactory(i),
+        stringFactory(i),
+        timestampFactory(i),
+        objectIdFactory(i),
+        uuidFactory(i),
+      ),
+    );
+    realm.write(() => realm.addAll(notIndexed));
+
+    expect(allNotIndexed.length, max);
+
+    // Inefficient, but fast enough for this test
+    final searchOrder = (List.generate(max, (i) => i)..shuffle(Random(42))).take(1000).toList();
```

Comment:
Why not

```suggestion
    final random = Random(42);
    final searchOrder = List.generate(1000, i => random.nextInt(max));
```

I know perf doesn't matter that much, it's just a little easier to read.

Reply:
Problem is it can create duplicates. That is what I avoid above

Reply:
It can, but do we care?